### PR TITLE
Add doc for LS-to-LS communication

### PR DESCRIPTION
**Pull request by karenzone**
First review of config from Jordan's video.
What about Windows?

### CODE REVIEW

File: docs/static/ls-ls-config.asciidoc [hunk 1]
```diff
@@ -0,0 +1,105 @@
+[[ls-to-ls]]
+=== Configuring Logstash-to-Logstash Communication
+
+In a complex network environment, you may need two Logstash machines talking to each other. That is, the output from the first Logstash machine becomes the input for the second Logstash machine. You may need this configuration if the data path crosses network or firewall boundaries, for example.
+
+NOTE: Don't complicate your configuration unnecessarily. If you don't have a compelling need for Logstash-to-Logstash communication, then don't implement it.
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: Do we have a way to format this as an aside? Otherwise, I'd just gently incorporate it into the preceding paragraph.

File: docs/static/ls-ls-config.asciidoc [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+[[ls-to-ls-overview]]
+==== Configuration overview
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: We could really use a diagram here. @acchen97 do you know where we can get one of those?

File: docs/static/ls-ls-config.asciidoc [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+. Use the Beats input plugin on the second Logstash machine.
+. Generate a trusted SSL certificate (required by the lumberjack protocol).
+. Save the SSL certificate on the first Logstash machine.
+. Save the SSL certificate and key on the second Logstash machine.
+. Test it.
+
+[[generate-self-signed-cert]]
+===== Generate a self-signed SSL certificate and key
+
+You can use the `openssl req` command with some specific flags to generate a self-signed certificate and key. The `openssl req` command is available on any operating system that comes with SSL.
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: I think we can strike `with some specific flags`

File: docs/static/ls-ls-config.asciidoc [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+[source,shell]
+----
+Generating a 2048 bit RSA private key
+.................................+++
+....................+++
+writing new private key to 'lumberjack.key'
+----
+
+[[save-cert-ls1]]
+===== Save the SSL certificate on the first Logstash machine
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: The heading indicates this section is about saving the certificate, but the command is about running the upstream logstash.

I think we want to amend the previous section to end by saying "then copy the generated key and certificate to all machines that will be communicating with each other". The section probably should be titled something like: "Trying it out" or "Testing it out"

File: docs/static/ls-ls-config.asciidoc [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+On the first Logstash machine, run the following command:
+[source,shell]
+----
+bin/logstash --path.data /tmp/z -e 'input { generator { count => 5 } } output { lumberjack { codec => json hosts => "localhost" ssl_certificate => "lumberjack.crt" port => 5000 } }'
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: We should explain why we're setting `path.data` here, which is to not have two logstashes attempt to use the same data path, which would prevent one from starting. A note saying that we're only doing this to be able to quickly demo this makes sense.

File: docs/static/ls-ls-config.asciidoc [hunk 6]
```diff
@@ -0,0 +1,105 @@
+[[ls-to-ls]]
+=== Configuring Logstash-to-Logstash Communication
+
+In a complex network environment, you may need two Logstash machines talking to each other. That is, the output from the first Logstash machine becomes the input for the second Logstash machine. You may need this configuration if the data path crosses network or firewall boundaries, for example.
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: I would add to this first section a succinct description of the solution, for those who don't need a tutorial. "The recommended way for one Logstash to communicate with another is by connecting the Lumberjack output to the Beats input", which links to the output and input respectively.

Comment (andrewvc) [reply docs/static/ls-ls-config.asciidoc]: Can we globally replace 'first' and 'second' with 'upstream' and 'downstream', that more clearly documents how these Logstash instances relate to one another.

I think we can generally reword this to say something along the lines of.

"To connect one Logstash to another we use the Lumberjack protocol, which is spoken by the lumberjack output and the beats input. To use it, setup the downstream Logstash instance with the Beats input to listen for incoming Lumberjack connections, then, setup the upstream Logstash instance with the Lumbjerjack output to connect to it."

File: docs/static/ls-ls-config.asciidoc (hunk 5 above)
Comment (karenzone) [reply docs/static/ls-ls-config.asciidoc]: Hmmmm... I meant to modify that sample based on the premise that this was a symptom of the demo and not likely in real world set up. In other words, I'm trying to help people get set up rather than helping them demo the solution.  1. Is that an accurate assumption? 2. Is setup rather than demo the correct approach for the reference guide?

File: docs/static/ls-ls-config.asciidoc [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+* `localhost` specifies the name of the first Logstash computer
+
+[[save-cert-ls2]]
+===== Save the SSL certificate and key on the second Logstash machine
+
+On the second Logstash machine, run the following command:
+
+[source,shell]
+----
+bin/logstash -e 'input {beats { codec =>? json port => 5000 ssl => ssl_certificate => "lumberjack.crt" ssl_key => "lumberjack.key"} }'
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: We can probably leave off the `codec` declaration.

File: docs/static/ls-ls-config.asciidoc [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+* `localhost` specifies the name of the upstream Logstash computer
+
+[[save-cert-ls2]]
+===== Save the SSL certificate and key on the downstream Logstash machine
+
+On the downstream Logstash machine, run the following command:
+
+[source,shell]
+----
+bin/logstash -e 'input {beats { codec =>? json port => 5000 ssl => ssl_certificate => "lumberjack.crt" ssl_key => "lumberjack.key"} }'
```
Comment (jordansissel) [inline docs/static/ls-ls-config.asciidoc]: is the question mark here intended? I don't' see it anywhere else: `codec =>? json`

**Comment by jordansissel** [comment]
LGTM

File: docs/static/ls-ls-config.asciidoc [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+[[generate-self-signed-cert]]
+===== Generate a self-signed SSL certificate and key
+
+You can use the `openssl req` command to generate a self-signed certificate and key. The `openssl req` command is available on any operating system that comes with SSL.
+
+Run the following command:
+
+[source,shell]
+----
+openssl req -x509 -batch -nodes -newkey rsa:2048 -keyout lumberjack.key -out lumberjack.crt -subj /CN=localhost
```
Comment (jordansissel) [inline docs/static/ls-ls-config.asciidoc]: Oops, found a typo.

Here we say `lumberjack.crt` but down below in `where:` we say `lumberjack.cert` (crt vs cert)

**Comment by jordansissel** [comment]
I forgot I could edit this PR, so I changed all the .crt command line references to be .cert to align with the rest of the doc. 

LGTM

**Comment by jordansissel** [comment]
@andrewvc I think this is pending your approval now.

File: docs/static/ls-ls-config.asciidoc [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+* `localhost` specifies the name of the upstream Logstash computer
+
+[[save-cert-ls2]]
+===== Save the SSL certificate and key on the downstream Logstash machine
+
+On the downstream Logstash machine, run the following command:
+
+[source,shell]
+----
+bin/logstash -e 'input {beats { codec =>? json port => 5000 ssl => ssl_certificate => "lumberjack.cert" ssl_key => "lumberjack.key"} }'
```
Comment (yaauie) [inline docs/static/ls-ls-config.asciidoc]: `codec =>? json` should probably be `codec => json`

Comment (yaauie) [inline docs/static/ls-ls-config.asciidoc]: `ssl => ssl_certificate => "lumberjack.cert"` should probably be `ssl_certificate => "lumberjack.cert"`

Comment (yaauie) [reply docs/static/ls-ls-config.asciidoc]: inconsistent spacing: `input {beats` should probably be `input { beats`.

File: docs/static/ls-ls-config.asciidoc [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+----
+
+where:
+
+* `lumberjack.key` specifies the name of the SSL key to be created
+* `lumberjack.cert` specifies the name of the SSL certificate to be created
+* `localhost` specifies the name of the computer
```
Comment (yaauie) [inline docs/static/ls-ls-config.asciidoc]: `localhost` isn't in the input config, since the downstream will be _listening_ and doesn't need to know the address of the upstream machine(s)

File: docs/static/ls-ls-config.asciidoc [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+On the downstream Logstash machine, run the following command:
+
+[source,shell]
+----
+bin/logstash -e 'input {beats { codec =>? json port => 5000 ssl => ssl_certificate => "lumberjack.cert" ssl_key => "lumberjack.key"} }'
+----
+
+where:
+
+* `lumberjack.key` specifies the name of the SSL key to be created
```
Comment (yaauie) [inline docs/static/ls-ls-config.asciidoc]: the key should already be created by the time we get here, so the "to be created" is misleading; this needs to be the _path_ to the key, not just the name.

File: docs/static/ls-ls-config.asciidoc [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+* `lumberjack.cert` specifies the name of the SSL certificate to be created
```
Comment (yaauie) [inline docs/static/ls-ls-config.asciidoc]: the cert should already be created by the time we get here, so the "to be created" is misleading; this needs to be the _path_ to the cert, not just the name.

File: docs/static/ls-ls-config.asciidoc [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+On the upstream Logstash machine, run the following command:
+[source,shell]
+----
+bin/logstash --path.data /tmp/z -e 'input { generator { count => 5 } } output { lumberjack { codec => json hosts => "localhost" ssl_certificate => "lumberjack.cert" port => 5000 } }'
+----
+
+where:
+
+* `5` specifies the number of events you want to generate and send
+* `lumberjack.cert` specifies the name of the SSL certificate you created
```
Comment (yaauie) [inline docs/static/ls-ls-config.asciidoc]: the cert should already be created by the time we get here, so the "to be created" is misleading; this needs to be the _path_ to the cert, not just the name.

File: docs/static/ls-ls-config.asciidoc [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+* `localhost` specifies the name of the upstream Logstash computer
```
Comment (yaauie) [inline docs/static/ls-ls-config.asciidoc]: "name" should probably be "address", and refers to the _downstream_ Logstash computer (a.k.a., where this output is going to send data _to_.

File: docs/static/ls-ls-config.asciidoc [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+On the upstream Logstash machine, run the following command:
+[source,shell]
+----
+bin/logstash --path.data /tmp/z -e 'input { generator { count => 5 } } output { lumberjack { codec => json hosts => "localhost" ssl_certificate => "lumberjack.cert" port => 5000 } }'
+----
+
+where:
+
+* `5` specifies the number of events you want to generate and send
```
Comment (yaauie) [inline docs/static/ls-ls-config.asciidoc]: may be helpful to include keys and sections in the breakdown, e.g.:
~~~ asciidoc
where:

* in the generator input:
** `count => 5` specifies the number of events you want to generate and send
* in the lumberjack output:
** `codec => json`: indicates that we should send fully-structured events
** `hosts => "localhost"`: indicates the network address to which we should send events
** `port => 5000`: indicates the port to which we send events 
** `ssl_certificate => "lumberjack.cert"`: indicates the local filesystem path to our certificate
~~~

File: docs/static/ls-ls-config.asciidoc [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+writing new private key to 'lumberjack.key'
+----
+
+[[save-cert-ls1]]
+===== Save the SSL certificate on the upstream Logstash machine
+
+On the upstream Logstash machine, run the following command:
+[source,shell]
+----
+bin/logstash --path.data /tmp/z -e 'input { generator { count => 5 } } output { lumberjack { codec => json hosts => "localhost" ssl_certificate => "lumberjack.cert" port => 5000 } }'
```
Comment (jordansissel) [inline docs/static/ls-ls-config.asciidoc]: Thinking about this more -- I'm not sure if we need to specify `--path.data /tmp/z` here. If the assumption is two independent servers running logstash, this flag (`--path.data`) is not necessary. It's only necessary when testing locally with two Logstash instances on the same machine (which is probably rare for this use case).

File: docs/static/ls-ls-config.asciidoc (hunk 10 above)
Comment (jordansissel) [reply docs/static/ls-ls-config.asciidoc]: This worked for me in testing.
```
bin/logstash -e 'input { beats { codec => json port => 5000 ssl => true ssl_certificate => "lumberjack.cert" ssl_key => "lumberjack.key" } }' 
```

File: docs/static/ls-ls-config.asciidoc (hunk 17 above)
Comment (jordansissel) [reply docs/static/ls-ls-config.asciidoc]: I tested this and was successful: 

```
  bin/logstash --path.data /tmp/z -e 'input { generator { count => 5 } } output { lumberjack { codec => json hosts => "localhost" ssl_certificate => "lumberjack.cert" port => 5000 } }'
```

👍

**Comment by jordansissel** [comment]
I ran the 3 commands (openssl, bin/logstash, bin/logstash) and with minor changes was successful -- I put some comments about the changes in the review. Let me know if you need anything other testing :)

File: docs/static/ls-ls-config.asciidoc [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+[[test-ls-to-ls]]
+===== Test the communication
+
+Use the input generator to send test events from the upstream Logstash machine.
+
+The following command sends five events over port 5000:
+
+[source,shell]
+----
+bin/logstash —path.data /tmp/z -e 'input { generator {count => 5 } } output { lumberjack { codec => json hosts => "localhost" ssl_certificate => "lumberjack.cert" Port => 5000 } }'
```
Comment (karenzone) [inline docs/static/ls-ls-config.asciidoc]: Rats.  I moved `—path.data /tmp/z -e` to the downstream instance in response to a review comment. So looks like I need to remove it here, too

File: docs/static/ls-ls-config.asciidoc [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+
+[[save-cert-key]]
+===== Save the SSL certificate and key
+
+Copy the SSL certificate to the upstream Logstash machine.
+
+Copy the SSL certificate and key to the downstream Logstash machine.
+
+[[config-demo]]
+==== Demo
```
Comment (karenzone) [inline docs/static/ls-ls-config.asciidoc]: In the first draft, i had a weird blend of configuration and demo.  I restructured the document in an attempt to fix that. Let me know what you think

File: docs/static/ls-ls-config.asciidoc [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+You can set up communication between two Logstash machines by connecting the Lumberjack output to the Beats input. You may need this configuration if the data path crosses network or firewall boundaries, for example. If you don't have a compelling need for Logstash-to-Logstash communication, then don't implement it.
+
+[[ls-to-ls-overview]]
+==== Configuration overview
+
+Use the Lumberjack protocol to connect two Logstash machines.
+
+. Generate a trusted SSL certificate (required by the lumberjack protocol).
+. Save the SSL certificate on the upstream Logstash machine.
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: "Save" is a weird verb for server side software, since there isn't really a 'save' action in the software operators use for the most part, and there's no unsaved state.

I think "copy" or "put" make more sense in general for Logstash and similar software.

File: docs/static/ls-ls-config.asciidoc [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+. Save the SSL certificate and key on the downstream Logstash machine.
+. Set the downstream Logstash machine to listen for incoming Lumberjack connections through the Beats input.
+. Set the upstream Logstash machine to use the Lumberjack output to send data.
+. Test it.
+
+[[generate-self-signed-cert]]
+===== Generate a self-signed SSL certificate and key
+
+You can use the `openssl req` command to generate a self-signed certificate and key. The `openssl req` command is available on any operating system that comes with SSL.
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: Not really true that the command is available on any operating system with SSL. We can replace the last sentence with: "You may need to install the openssl command line program".

File: docs/static/ls-ls-config.asciidoc (hunk 12 above)
Comment (andrewvc) [reply docs/static/ls-ls-config.asciidoc]: @yaauie +1 on path vs. name. However, I think 'to be created' is correct. This describes what the `openssl` command does with the argument, which is, in fact, the creation of the key file.

File: docs/static/ls-ls-config.asciidoc (hunk 13 above)
Comment (andrewvc) [reply docs/static/ls-ls-config.asciidoc]: Same thought as above, we're documenting the 'openssl' command, not where we are at this point in time.

File: docs/static/ls-ls-config.asciidoc [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+Generating a 2048 bit RSA private key
+.................................+++
+....................+++
+writing new private key to 'lumberjack.key'
+----
+
+[[save-cert-key]]
+===== Save the SSL certificate and key
+
+Copy the SSL certificate to the upstream Logstash machine.
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: Yes! I much prefer the use of copy here to save :)

File: docs/static/ls-ls-config.asciidoc [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+Run the following command to start Logstash and generate test events:
+
+[source,shell]
+----
+bin/logstash -e 'input { generator { count => 5 } } output { lumberjack { codec => json hosts => "localhost" ssl_certificate => "lumberjack.cert" port => 5000 } }'
+----
+
+where:
+
+* `5` specifies the number of events to be generated
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: If we do keep this, as a convention should we use `5` or `count => 5` or `count`? In other words, do we document the value or the key for that value? This is something we should decide as a style convention for our docs in general.

File: docs/static/ls-ls-config.asciidoc [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+===== Start the upstream Logstash instance
+
+Run the following command to start Logstash and generate test events:
+
+[source,shell]
+----
+bin/logstash -e 'input { generator { count => 5 } } output { lumberjack { codec => json hosts => "localhost" ssl_certificate => "lumberjack.cert" port => 5000 } }'
+----
+
+where:
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: I almost wonder if we don't need so much structure here. This could be a paragraph. "This command will send five events to localhost:5000 using the provided SSL certificate to verify the connection".

We don't need to document everything, since this page only is about LS->LS communication

File: docs/static/ls-ls-config.asciidoc [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+Run the following command to start the downstream instance of Logstash:
+
+[source,shell]
+----
+bin/logstash --path.data /tmp/z -e 'input { beats { codec => json port => 5000 ssl_certificate => "lumberjack.cert" ssl_key => "lumberjack.key"} }'
+----
+
+where:
+
+* `--path.data /tmp/z` sets a different data path (required to start a second Logstash instance on the same computer)
```
Comment (andrewvc) [inline docs/static/ls-ls-config.asciidoc]: Same note here, do we want a list or a paragraph?

File: docs/static/ls-ls-config.asciidoc (hunk 24 above)
Comment (andrewvc) [reply docs/static/ls-ls-config.asciidoc]: On the other hand, for some users, this info might be valuable. I am worried about us re-documenting how every LS option works on every page.

Comment (karenzone) [reply docs/static/ls-ls-config.asciidoc]: @dedemorton I'd like to get your thoughts on the best approach

File: docs/static/ls-ls-config.asciidoc (hunk 23 above)
Comment (karenzone) [reply docs/static/ls-ls-config.asciidoc]: @dedemorton I'd like to get your thoughts on the best approach. (I like to use a sample value and then explaining it rather than going too generic. At the same time, I want to be consistent with other writers and products.)

Comment (dedemorton) [reply docs/static/ls-ls-config.asciidoc]: @karenzone The page was not refreshed, so I didn't see your comment until after I wrote mine. Striving for consistency is a bit tricky atm because we have a lot of hands in the docs. I'd focus on making your content internally consistent, accurate, and easy-to-read, and not worry as much about other projects and writers.
